src/cruise_control/scripts/cruise_control.py

```diff
@@ -5,34 +5,50 @@
 from sensor_msgs.msg import PointCloud
 from std_msgs.msg import Float64
 
-
 class CruiseControl:
 
     def __init__(self):
         self.steeringPub = rospy.Publisher("steering", Float64, queue_size = 10)
         self.brakingPub = rospy.Publisher("braking", Float64, queue_size = 10)
         self.throttlePub = rospy.Publisher("throttling", Float64, queue_size = 10)
+        self.speed = 0.0
 
     def listener(self):
         rospy.init_node("cruise_control", anonymous=True)
         rospy.Subscriber("lidar", PointCloud, self.handle_lidar_data)
         # Define Path data structure
         rospy.Subscriber("pathData", PathData, self.handle_path_data)
-        rospy.spin()
+        rospy.Subscriber("sensor/speed", Float64, self.handle_speed_data)
+
+        # Node is publisher and subscriber- cannot use spin; the publisher methods will never get called
+
+        # Probably should have same rate for this node and speedometer to make them more in sync
+        # Higher the rate the smoother the constant speed is, but will take more CPU power
+        rate = rospy.Rate(30)
+        while not rospy.is_shutdown():
+         self.publish_results()
+         rate.sleep()
 
     def handle_lidar_data(self, data):
         print("Obtained lidar data")
 
     def handle_path_data(self, data):
         print("Obtained path data")
 
+    def handle_speed_data(self, speed: Float64):
+        rospy.loginfo("Handling the speed data")
+        self.speed = speed.data
+
     def publish_results(self):
-        rate = rospy.Rate(2)
-        while rospy.is_shutdown():
-            self.steeringPub.publish(1)
-            self.brakingPub.publish(1)
-            self.throttlePub.publish(1)
-            rate.sleep()
+        self.steeringPub.publish(1)
+        self.brakingPub.publish(1)
+
+        if self.speed < 5:
+            # If throttle is too low then car will not go above a certain speed (at least if road is
+            # uphill). This value might have to change as a result
```

Comment:
More specifically, given the starting default position, a throttle of 0.25 makes the car never go above 2 m/s

Reply:
Just to make sure that 5 is the max speed value for the car?

Reply:
Yeah, though depending on granularity of the rate of the node and the throttle value, it will temporarily go above this speed before the node realizes the car is going too fast and then reduce throttle